src/types/string.rs

```diff
@@ -56,18 +56,14 @@ impl PyString {
         }
     }
 
-    /// Gets the python string data in its underlying representation.
-    pub fn data(&self) -> PyStringData {
-        // TODO: return the original representation instead
-        // of forcing the UTF-8 representation to be created.
+    /// Get the Python string as a byte slice.
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
         unsafe {
             let mut size: ffi::Py_ssize_t = mem::uninitialized();
             let data = ffi::PyUnicode_AsUTF8AndSize(self.0.as_ptr(), &mut size) as *const u8;
-            if data.is_null() {
-                PyErr::fetch(self.py()).print(self.py());
-                panic!("PyUnicode_AsUTF8AndSize failed");
-            }
-            PyStringData::Utf8(std::slice::from_raw_parts(data, size as usize))
+            debug_assert!(!data.is_null());
```

Comment:
Ok, added.

Reply:
Looking at the [original c source](https://github.com/python/cpython/blob/e2f48bf0e851dc79d888c70fd24c631a8a9e1030/Objects/unicodeobject.c#L3740-L3773), there seems to be a cases where even a valid PyObject would return null:

```c
const char *
PyUnicode_AsUTF8AndSize(PyObject *unicode, Py_ssize_t *psize)
{
    PyObject *bytes;


    if (!PyUnicode_Check(unicode)) {
        PyErr_BadArgument();
        return NULL;
    }
    if (PyUnicode_READY(unicode) == -1)
        return NULL;


    if (PyUnicode_UTF8(unicode) == NULL) {
        assert(!PyUnicode_IS_COMPACT_ASCII(unicode));
        bytes = _PyUnicode_AsUTF8String(unicode, NULL);
        if (bytes == NULL)
            return NULL;
        _PyUnicode_UTF8(unicode) = PyObject_MALLOC(PyBytes_GET_SIZE(bytes) + 1);
        if (_PyUnicode_UTF8(unicode) == NULL) {
            PyErr_NoMemory();
            Py_DECREF(bytes);
            return NULL;
        }
        _PyUnicode_UTF8_LENGTH(unicode) = PyBytes_GET_SIZE(bytes);
        memcpy(_PyUnicode_UTF8(unicode),
                  PyBytes_AS_STRING(bytes),
                  _PyUnicode_UTF8_LENGTH(unicode) + 1);
        Py_DECREF(bytes);
    }


    if (psize)
        *psize = PyUnicode_UTF8_LENGTH(unicode);
    return PyUnicode_UTF8(unicode);
}
``` 

Imo we don't need to care about stuff like out of memory, but we should note that this only works because we now that we have a `PyString`/`PyUnicode`

Reply:
Ok, thanks for looking at that. I've updated it.